### PR TITLE
Animated webp

### PR DESCRIPTION
**Pull request by eylles**
now that it has been documented this is a purely optional thing i see no reason not to merge it.

### CODE REVIEW

File: Makefile
Comment (N-R-K) [inline Makefile:35]: If https://github.com/nsxiv/nsxiv/pull/19 is to be merged, this will need some rework.

File: image.c
Comment (N-R-K) [inline image.c:148]: What's going on here? I don't see any reason to change indentation. There's a couple more places where indentation was changed.

I think we should keep indentation as is, if we are to change it, it should be changed throughout the codebase and should be its own commit/PR.

**Review by N-R-K — CHANGES_REQUESTED** [review]
I haven't looked into the `img_load_webp` function yet. As of now here's some basic feedback...

File: image.c
Comment (eylles) [reply image.c:148]: that was done by ytrew, i opened this on a "as is" state to get feedback and correct these kind of things, will go through it later and make sure the indentation is consistent.

File: Makefile
Comment (eylles) [reply Makefile:35]: well, it will depend on which PR gets merged first, if this gets merged first then #19 will need a rebase.

File: image.c
Comment (N-R-K) [inline image.c:327]: I don't believe this is needed. `check_add_file` already checks for this.

File: image.c
Comment (N-R-K) [inline image.c:427]: Once again, `check_add_file` already checks for this. Also why should we disallow loading a webp file if the extension isn't `.webp`.

It's perfectly fine to load an image named `img` without any extension.

**Review by N-R-K — CHANGES_REQUESTED** [review]
Initial impression is that `img_load_webp()` and `load_webp_firstframe_im()` shares a massive amount of duplicate code. ~~I think we can merge them into one function and add a boolean arg `first_frame`.~~ Nvm, they have different return types. Still think there should room for refactoring.

Anyways, I'll delve more into this tomorrow. Also I'm not familiar with libwebp so I'll need to look up the docs. As of now, I believe this patch is too messy to be merged.

File: image.c
Comment (eylles) [reply image.c:327]: will start tests on removing this and checking that nothing breaks, in theory it should not break as this is redundant.

File: image.c
Comment (N-R-K) [inline image.c:43]: One of the function is declared here, but the other isn't.

File: image.c
Comment (N-R-K) [inline image.c:554]: This seems to to be the only place where this function is called. And the return value isn't even used.

**Review by N-R-K — CHANGES_REQUESTED** [review]
The entire patch pretty much needs to be rethought and refactored. I'll give it a go tomorrow. In the meanwhile if anyone has any ideas on how to go about doing so, please share them.

File: image.c
Comment (eylles) [reply image.c:327]: doing some tests this is certainly redundant and can be removed without any problem

File: image.c
Comment (eylles) [reply image.c:427]: after some tests the patch has a general problem with images without extensions, trying to load any image without extension, not just webp (static or animated) will cause a segfault.

Comment (eylles) [reply image.c:427]: after further testing this very check is the culprit of sxiv segfaulting on every extensionless image, without it a extensionless animated webp will load and play just fine without any problem, however the stdout will be littered by a `sxiv: /path/to/non_webp.img: Error opening webp image` message on every non webp image

**Comment by eylles** [comment]
will change this to a draft while refactoring is done, in the meantime i will continue making some more tests to try and narrow down places where things go south.

File: image.c
Comment (eylles) [inline image.c:474]: after some testing i was able to determine that with the previously identified redundant checks removed it is this if statement that will constantly spew out the error message on every non webp image and on extensionless webp images, however simply removing this check will cause a segfault when reading large .jpg images.

**Comment by bakkeby** [comment]
I think for a pull request like this that we should squash commits rather than having 7 individual changes.

sxiv has support for the webp image format out of the box through the imlib2 library and this is a quite substantial amount of code to add just to get animation working.

The ugly part of this is that you need to try to open an image (.png, jpg, etc.) as a webp image first to see if it is an animated webp file, then if it is not fallback to opening the file as normal via imlib2. I am not sure if there is really a good way around this.

The underlying issue here being that imlib2 does not support animation, which is also why we have separate code to deal with playing animated gifs. To what degree imlib2 is still maintained I am not sure, but I doubt that animation is something that will be added in the foreseeable future.

Besides it being cool to show off to the ladies - do anyone actually use / depend on support for animated web images? Despite being created by Google it still seems like a bit of a niche thing?

**Comment by N-R-K** [comment]
I pretty much agree with you on all the points, however:

> sxiv has support for the webp image format out of the box through the imlib2 library

imblib2 depends on libwebp for that. So since libwebp will be pulled in, and we have a working (although EXTREMELY messy) prototype, I think we should work on refactoring it.

This PR shouldn't be a priority however, since animated webp is a very niche thing. I said I'd delve into this, but because of the reasons described above, I haven't yet.

Even if no one else is willing to work on it (understandable, since animated webp is rare), I think I can get it refactored later sometime... once more higher priority things have been dealt with.

**Comment by Ytrewq13** [comment]
I'm going to work on refactoring this PR today. I'll try to merge the duplicated code from `img_load_webp()` and `load_webp_firstframe_im()` into a shared function that both of those can call, with an argument to specify whether or not we want to populate the multiframe part of the `img_t`

> I think for a pull request like this that we should squash commits rather than having 7 individual changes.

I agree with this. The only reason I didn't do this is because I wasn't familiar with any of git's more advanced features at the time. I'll try to fix it if that's something I can do to the existing PR.

> The ugly part of this is that you need to try to open an image (.png, jpg, etc.) as a webp image first to see if it is an animated webp file, then if it is not fallback to opening the file as normal via imlib2. I am not sure if there is really a good way around this.

I'm trying to figure this out at the moment and I'll hopefully have a more elegant solution by tomorrow.

> Besides it being cool to show off to the ladies - do anyone actually use / depend on support for animated web images? Despite being created by Google it still seems like a bit of a niche thing?

The reason I decided to implement this was because animated WebP images are very often much smaller in file size than equivalent GIFs. I don't remember what GIFs I was looking at converting to WebPs at first, and I'll admit that it is very niche. I've never encountered an animated WebP image in the wild, and I had to create my own using GIMP and ImageMagick in order to test this patch.

I think this PR shouldn't be merged unless the result of refactoring it meshes very nicely with the rest of nsxiv's codebase. We don't want to introduce bloat for a feature that only 3 or 4 people have ever requested. A patch that users can choose to install would be more inline with suckless philosophy.

File: image.c
Comment (Ytrewq13) [reply image.c:43]: This was because `load_webp_firstframe()` is not used outside `image.c`, but `img_load_webp()` is called in `sxiv.c`.

I declared functions only in the scopes in which they were required, but I can understand the confusion. Do you think this should be done differently?

Comment (N-R-K) [reply image.c:43]: Aren't we planning on merging the two functions into one? This shouldn't matter in that case.

File: image.c
Comment (Ytrewq13) [reply image.c:554]: The function is implemented following `img_load_gif()`.

I don't think only calling it in one place is a big issue, as having that logic in its own function makes sense considering that `img_load()` is high-level and abstracted from the details of any image format.

I agree that the return value not being used is a problem, but the same problem exists for `img_load_gif()`, so I think the problem lies witht the implementation of the `img_load()` function. It shouldn't ignore the return values of `img_load_gif()` or `img_load_webp()`, and should instead check if they returned true before continuing with the rest of the function, bailing out and returning false if not.

**Comment by Ytrewq13** [comment]
> take the chunks of shared/duplicate code and put them into their own functions

Apologies, this is what I meant I waas going to do.
`img_load_webp()` and `load_webp_firstframe_im()` will (probably) both still exist when I'm done today, but they won't contain the logic for loading a frame from a WebP image. That will be removed into a new function which they will both use.

File: image.c
Comment (N-R-K) [reply image.c:554]: > I don't think only calling it in one place is a big issue

Apologies for being misleading. My point was that since the two functions had different return types, I thought that they couldn't be merged, but then I see that this function's return value isn't even used, so I'm thinking, "I guess it can be merged (?)"

**Comment by eylles** [comment]
>
>I think this PR shouldn't be merged unless the result of refactoring it meshes very nicely with the rest of nsxiv's codebase. We don't want to introduce bloat for a feature that only 3 or 4 people have ever requested. A patch that users can choose to install would be more inline with suckless philosophy.

On this cSe adding the support for this is more of case of "for completeness" as the underlaying library used by sxiv, imlib2, optionally supports static webp then it makes sense sciv should optiobally support the animated version, which by that logic means we may have to look into supporting apng in the future (definitely outside the scope of the first release of nsxiv, and probably outside the scope of the second release too)

**Comment by Ytrewq13** [comment]
> On this case adding the support for this is more of case of "for completeness" as the underlaying library used by sxiv, imlib2, optionally supports static webp then it makes sense sciv should optiobally support the animated version, which by that logic means we may have to look into supporting apng in the future

Since I spent so much time learning about WebP  (considering how much of a failure WebP was, I might be the person who knows it best aside from the team that developed libwebp), I would probably be well-positioned to take a look at apng. I have a good knowledge of file formats and I wouldn't mind learning about another multiframe image format.

I know we are a way off from adding support for apng, but I am volunteering myself for that task. Taking a very basic look at the spec, it seems that most of the ideas carry across from how WebP files are arranged, so it should be easy to pick up. Once this PR is refactored and merged, I'll put my spare time towards apng.

File: image.c
Comment (Ytrewq13) [reply image.c:43]: The way I'm refactoring it, both of these functions will still exist, but I'll create a new function which handles all of the shared code. Then the two existing functions would mostly just be wrappers for the new one with specific single-frame or multi-frame code.

It might be the case that the resulting functions are so trivial that I can then remove one or both of them and fully merge them into one function. Probably I'll leave `img_load_webp()` as a wrapper for a function that takes more arguments and can either load the first frame or all frames of a WebP image into an array of imlib2 images.

Basically, in response to your comment, I'll probably merge the functions into one, but that depends on how nicely the functions behave once I finish extracting all the shared code.

File: image.c
Comment (Ytrewq13) [reply image.c:474]: > large .jpg images.

How large exactly? I'm testing a potential fix which is on the way towards tidying up the patch and I'm not able to reproduce this issue.

Simply removing this check, I get a segfault reading any image with no extension, but it can read a 69MB .jpg image with no problem.

Comment (eylles) [reply image.c:474]: i gotta admit that one was quite inconsistent as i could read a 8MB .jpg just fine but a 19MB .jpg just next to it made it segfault, so not truly sure it was exactly that...

**Comment by Ytrewq13** [comment]
As an update for how the refactoring is going, here's what I've done so far:
- I've created a function that checks if any file is a valid WebP file. This function reads as little of the file as necessary so it is better than calling `load_webp_firstframe_im()` on every extensionless file.
- I've rearranged how `img_open()` is structured so we don't have two copies of the main format check. This rearranging has fixed issues with extensionless webp files and also fixed the `sxiv: filename: Error opening webp image` spam.

Here's what I still have left to do:
- Move all the shared code from `img_load_webp()` and `load_webp_firstframe_im()` into a new function, `Imlib_Image *load_webp_frames(const fileinfo_t *file, bool all_frames)`. If the boolean argument is false, the function will load only the first frame of the image, which is what `load_webp_firstframe_im()` does at the moment. If it is true, we will load every frame.
  - I will remove `load_webp_firstframe_im()` and only use the new function in its place. It's only called once, and the return value is similar enough (I just have to dereference what is returned in order to swap the functions).
  - Change the implementation of `img_load_webp()` to use the new function. `img_load_webp()` will only extract the frames that are loaded and populate the `img_t`.

**Comment by N-R-K** [comment]
@Ytrewq13 consider pushing your changes here so we can review.

File: image.c
Comment (Ytrewq13) [inline image.c:377]: I just realised I forgot to revert these debugging changes. Before the PR is merged I'm going to change the generic error messages to be a bit more specific, so that if there's a problem with the WebP library version the error message will mention that, instead of just saying "error opening webp image".

File: image.c
Comment (N-R-K) [inline image.c:310]: This doesn't seem like a high enough size worth mallocing.

**Review by N-R-K — CHANGES_REQUESTED** [review]
Didn't look too closely, but I like the direction it's taking. The newer `img_open` is much more sane.

File: image.c
Comment (N-R-K) [inline image.c:339]: All declarations should go on top of the block.

**Comment by N-R-K** [comment]
@Ytrewq13 I don't think we're using stdint types on this codebase. But since `WebPGetInfo` is expecting that, we can make an exception.

**Comment by N-R-K** [comment]
Did a shallow check and some cleanup. I say shallow because I'm not familiar with libwebp. However couldn't find any style problems anymore, so it seems that you've been accustomed to codebase a bit more now, which is good.

Anyways, when a webp is looping and we go from the last frame -> first frame, my cpu usage spikes up to 10-20%. I think we might be loading/looping things inefficiently. Can you take a look on whats causing this? @Ytrewq13

**Review by Ytrewq13 — COMMENTED** [review]
> when a webp is looping and we go from the last frame -> first frame

Does this definitely only happen when we go from the last to the first frame? I think that would be an issue with the animation code in `main.c`, maybe the `animate()` function at `main.c:447` or one of the functions called by that (`img_frame_animate()`?).

File: image.c
Comment (eylles) [reply image.c:321]: no point on trying to read an empty image i guess so this should be okay.

**Comment by N-R-K** [comment]
Went a bit aggressive on https://github.com/nsxiv/nsxiv/pull/20/commits/dadce942f6774fd1f190a6bec70fbdf54c5cae0a , I'm currently writing a function `is_gif` due to https://github.com/nsxiv/nsxiv/issues/37 and wanted both `is_gif` and `is_webp` to be similar in style.

Do you think this is fine or did I go too aggressive? @Ytrewq13

**Comment by Ytrewq13** [comment]
> Do you think this is fine or did I go too aggressive? @Ytrewq13

It looks ok to me. I'd even say you've improved it, we don't have to check the file size when we can just try to read what we need.

**Comment by Ytrewq13** [comment]
I have now fixed memory leaks in this PR so I think we're ready for final reviews, squashing the commits, and merging.

**Comment by N-R-K** [comment]
Thanks. But this pr will need to be properly reviewed. I cannot do that atm, since I'm not familiar with libwebp. And I cannot approve code which I don't properly understand. So we'll have to wait until someone takes a thorough look at the webp loading function.

Also we'll have to rebase this before merging and update the Makefile, since dev branch went ahead by quite a bit. I'll do that later, after the function has been thoroughly reviewed and approved.

**Comment by Ytrewq13** [comment]
> properly reviewed.

This is what I meant by final reviews. I was just saying that I'm now happy with the state of this feature.

**Comment by eylles** [comment]
welp, now it should be up to master, will be testing it for a while and see if i find some problem

**Comment by bakkeby** [comment]
I had a look at the complete diff and I must say it looks very good. Didn't run into any issues in my testing.

Kudos for also addressing the issue of opening extensionless webp images.

**Comment by N-R-K** [comment]
> Kudos for also addressing the issue of opening extensionless webp images.

The extension-less problem was `imlib2` regression, now fixed in git. See https://github.com/nsxiv/nsxiv/issues/37#issuecomment-920759273

And yes, I agree. Most of the low hanging fruits have been picked and the patch works much more sanely now. I will try to properly review this within this week and hopefully we can push this into the next release.

**Comment by N-R-K** [comment]
@Ytrewq13 cleaned up some stuff. Please check and confirm if there's anything wrong here.
Structure wise, I think we're doing okay now.
Next step for me is to look into the libwebp docs and review this thoroughly. I think we should be able to merge this into the next release.

**Review by Ytrewq13 — APPROVED** [review]

**Review by N-R-K — APPROVED** [review]

**Comment by N-R-K** [comment]
We're good here. However since @Ytrewq13 is the main author and I spent a good time refactoring it, *there might be something obvious that we missed.*

After one more maintainer has taken a fresh look at this PR and confirms everything is okay, we can proceed with merge. @kdkasad @bakkeby

**Comment by N-R-K** [comment]
I'm not sure if it's placebo but I think the playback slows down a bit when heavily zoomed in. However that's not really related to this PR but rather seems to be an `sxiv` bug itself. See: https://github.com/nsxiv/nsxiv/issues/70

**Review by Ytrewq13 — APPROVED** [review]

File: image.c
Comment (bakkeby) [inline image.c:406]: This is of course making assumptions that `imlib_image_set_has_alpha` supports values other than 0 and 1.

Comment (N-R-K) [reply image.c:406]: The document only mentions `0` and `1`, however the function simply does an if check on the param. So it should work.
But I guess it's best to be safe, I'll update this.

**Review by N-R-K — CHANGES_REQUESTED** [review]
@Ytrewq13 Hmm, found a bug. When enabling alpha via <kbd>A</kbd> and then going to thumbnail mode and back to normal mode, the alpha background gets reset back to default window bg. With other img formats, this doesn't occur.

**Comment by Ytrewq13** [comment]
> @Ytrewq13 Hmm, found a bug. When enabling alpha via A and then going to thumbnail mode and back to normal mode, the alpha background gets reset back to default window bg. With other img formats, this doesn't occur.

This was caused by `img_load_webp()` setting `img->alpha` equal to `ALPHA_LAYER`, which is the default setting for whether the alpha layer is displayed(?). I removed that line and it fixed the bug.

We should probably design a series of tests to perform for future changes that would identify bugs like this.

**Review by N-R-K — APPROVED** [review]

**Comment by N-R-K** [comment]
Given that the only concern bakkeby had has been addressed and no one else seems to have any problem with it, I think we should be good to merge.

**Pull request merged**
